libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+		// This shouldn't be a fatal error, as this can only be called once per thread.
+		// Specifically, this will return an error in unit tests.
+		log.Warn("firewood: error starting logs", "error", err)
+	}
+
+	dbPath := filepath.Join(config.ChainDir, dbFileName)
+	fw, err := ffi.New(dbPath, &ffi.Config{
+		NodeCacheEntries:     uint(config.CleanCacheSize) / 256, // TODO: estimate 256 bytes per node
+		FreeListCacheEntries: config.FreeListCacheEntries,
+		Revisions:            config.Revisions,
+		ReadCacheStrategy:    config.ReadCacheStrategy,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	currentRoot, err := fw.Root()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		Firewood:    fw,
+		proposalMap: make(map[common.Hash][]*proposal),
+		proposalTree: &proposal{
+			info: &blockInfo{
+				root: common.Hash(currentRoot),
+				hashes: map[common.Hash]struct{}{
+					{}: {}, // genesis block has empty hash
+				},
+			},
+		},
+	}, nil
+}
+
+func validatePath(dir string) error {
+	if dir == "" {
+		return errors.New("chain data directory must be set")
+	}
+
+	// Check that the directory exists
+	switch info, err := os.Stat(dir); {
+	case os.IsNotExist(err):
+		log.Info("Database directory not found, creating", "path", dir)
+		if err := os.MkdirAll(dir, 0o755); err != nil {
+			return fmt.Errorf("error creating database directory: %w", err)
+		}
+		return nil
+	case err != nil:
+		return fmt.Errorf("error checking database directory: %w", err)
+	case !info.IsDir():
+		return fmt.Errorf("database directory path is not a directory: %s", dir)
+	}
+
+	return nil
+}
+
+// Scheme returns the scheme of the database.
+// This is only used in some API calls
+// and in StateDB to avoid iterating through deleted storage tries.
+// WARNING: If anything from go-ethereum uses this,
+// it must be overwritten to use something like:
+// `_, ok := db.(*Database); if !ok { return "" }`
+// to recognize the Firewood database.
+func (*Database) Scheme() string {
+	return rawdb.HashScheme
+}
+
+// Initialized checks whether a non-empty genesis block has been written.
+func (db *Database) Initialized(common.Hash) bool {
+	rootBytes, err := db.Firewood.Root()
+	if err != nil {
+		log.Error("firewood: error getting current root", "error", err)
+		return false
+	}
+	root := common.BytesToHash(rootBytes)
+	// If the current root isn't empty, then unless the database is empty, we have a genesis block recorded.
+	return root != types.EmptyRootHash
+}
+
+// Size returns the storage size of diff layer nodes above the persistent disk
+// layer and the dirty nodes buffered within the disk layer
+// Only used for metrics and Commit intervals in APIs.
+// This will be implemented in the firewood database eventually.
+// Currently, Firewood stores all revisions in disk and proposals in memory.
+func (*Database) Size() (common.StorageSize, common.StorageSize) {
+	return 0, 0
+}
+
+// This isn't called anywhere in coreth
```

Comment:
`libevm` is agnostic to `coreth` and `subnet-evm` (and all consumers for that matter).

```suggestion
// Reference is a no-op because...
```

Why doesn't Firewood need referencing? Because of the finalizer semantics we're working on?

Is `Reference` typically used as a keepalive signal?

Reply:
Oops, yeah that's outdated. From triedb docs:
```
// Reference adds a new reference from a parent node to a child node. This function
// is used to add reference between internal trie node and external node(e.g. storage
// trie root), all internal trie nodes are referenced together by database itself.
```
It's only used for HashDB anyway. But yes, it's only exported as a keep-alive signal

Reply:
I did update the comment